### DISCUSSION TITLE
Constraint Handling in MC Acquisition Functions

### DISCUSSION THREAD

**Discussion by polivucci** [meta-pytorch/botorch#2292] [General]
Hi all,

First off, thanks to the developers for the fantastic work on BOTorch.

I would like to make sense of constraint handling for MC Acquisition Functions (AFs).
As far as I know, BOTorch implements 3 ways of handling constraints:
1) Through the feasibility-weighted objective `ConstrainedMCObjective()`. Here , the AF is evaluated on a objective which has been already weighted with a feasibility indicator $FI$ as: $\alpha(g(x)FI(x))$
2) Through the `SampleReducingMCAcquisitionFunction()` AF class.  Here feasibility weighting is introduced after evaluating the AF: $\alpha(g(x))FI(x)$ This seems similar to the analytic Constrained Expected Improvement.
3)  Via `ConstrainedMaxPosteriorSampling`. This is a constrained version of Thompson Sampling from the SCBO paper.

My understanding is that the applicability of the above goes as follows:
1) is applicable to all MC acquisition functions of the `MCAcquisitionFunction()` class , to which we can pass `ConstrainedMCObjective()` through the `objective` argument.  This includes most MC AFs with the exception of Entropy Search acquisitions which define their own AF base class.
2) is ofc applicable to all MC acquisition functions of the `SampleReducingMCAcquisitionFunction()` class, which as of now only include the EI and logEI family.
3) is only specific to Thompson Sampling. However, I believe passing an instance of `ConstrainedMCObjective()` to `MaxPosteriorSampling` would also work out as it admits an `objective` argument.

Therefore it seems to me that `ConstrainedMCObjective()` is the only method common to almost all MC AFs.
Exception is Entropy Search AFs for which constraining is atm not yet supported (unless perhaps one can make it work via posterior transform?).

Is my understading correct? 
It looks to me that approaches 1. and 2. are not equivalent, so I would like to ask if and how they differ in practice and if you recommend one over the other, based on your experience. Intuitively I would lean towards 2. wherever possible.
Likewise for TS, SCBO handling vs the constrained objective.

Many thanks for your help!

**Comment by sdaulton** [comment]
Hi @polivucci, That's right. 

2) is the preferred approach since it applies constraints on a sample level to the sample-level utility (improvement) for each individual candidate in the batch. 2 is a smooth relaxation of that approach in [Letham et al, 2019](https://projecteuclid.org/journals/bayesian-analysis/volume-14/issue-2/Constrained-Bayesian-Optimization-with-Noisy-Experiments/10.1214/18-BA1110.full). This is also the approach used in multi-objective acquisition functions.

1) is a different approach, where as you mentioned, the feasibility indicator is applied to samples of $f$ rather than to the sample-level utility. This is not an equivalent approach and can lead to more harshly penalize points that are not likely to be feasible. 

@SebastianAment has done a deep dive on this and may have more thoughts.

**Reply by SebastianAment** [reply]
Hi @polivucci,

I second @sdaulton's comment that 2. is the preferred approach. Adding a few more technical details:
- Option 1. is likely to give rise to numerical problems. First, the objective function needs to be shifted to be positive before applying the feasibility indicator: $(g(x) - \min_z g(z)) FI(x) + \min_z g(z)$. This is done automatically using [`get_infeasible_cost`](https://github.com/pytorch/botorch/blob/e8cbbaec8f4d94af18c33aca271ff4265504339f/botorch/acquisition/utils.py#L203). The main issue arises since the weighted objective values can be close to zero in a large portion of the space, depending how difficult it is to satisfy the constraints, leading to numerically vanishing gradients that can leave the acquisition function optimizer stuck in clearly infeasible regions. 
- Option 2. mitigates the vanishing gradient problem by transforming the improvement computations to a smoothed log space, and adding the log constraint indicator function ([botorch.utils.sigmoid](https://github.com/pytorch/botorch/blob/e8cbbaec8f4d94af18c33aca271ff4265504339f/botorch/utils/safe_math.py#L481C5-L481C12)), rather than multiplying with a potentially numerically vanishing number, i.e. $\log \min(g(x) - g^*) + \log(FI(x))$.

We saw particularly significant performance improvements using the smoothed log-transformed computations of LogEI for constrained problems, see Figure 4 of [Unexpected Improvements to Expected Improvement for Bayesian Optimization](https://arxiv.org/pdf/2310.20708.pdf).

Let me know if you have additional questions about the approach.

**Reply by polivucci** [reply]
Hi @SebastianAment, @sdaulton,

Many thanks for your kind replies!

I have a couple follow-up questions:

- My understanding is that using 2. is advatageous both in principle (as mentioned by @sdaulton) and numerically (due to log treatment @SebastianAment). Is this a correct view?

- I insisted on `ConstrainedMCObjective()` as it is to the best of my knowledge the only way to constrain single-objective AFs outside the EI family. I think many people (including myself) would like to test several AFs on their problem while maintaining a consistent constraint treatment.  Do you see approach 2. being extended to KG, ES and TS in the future? 

- I noticed that one can pass a series of exact input constraints to optimize_acqf(). I admittedly did not yet dig deep into the details, apologies if my questions sounds too vague. At the moment the way I prescribe exact nonlinear constraints on the inputs is by using a  GenericDeterministicModel() as an outcome constraint model. What is your recommended treatment of nonlinear input constraints? 

Thank you

**Reply by sdaulton** [reply]
> Is this a correct view?

yes

>Do you see approach 2. being extended to KG, ES and TS in the future?

KG doesn't support constraints (no published work on it). Constrained ES methods are more custom then just feasibility weighting the objective (any do not rely on MC sampling in the same way as an improvement-based AF). TS can easily be constrained by simply drawing a sample of the objective and constraint and computing the exact feasibility weighted objective under the sample (no smooth relaxation needed).

> What is your recommended treatment of nonlinear input constraints?

What you described would be my recommendation.

**Comment by polivucci** [comment]
Hi @sdaulton,

Thanks for your advice.

Regarding constrained KG:
- `qKnowledgeGradient` can run on `ConstrainedMCObjective`. Is there a reason for why doing so is not recommended in principle? Aside from potential AF vanishing gradients. For example, in the case of EI, [Letham et al, 2019](https://projecteuclid.org/journals/bayesian-analysis/volume-14/issue-2/Constrained-Bayesian-Optimization-with-Noisy-Experiments/10.1214/18-BA1110.full) say their formulation is superior to another when observation noise is present.
- what is your opinion on  [Ungredda and Branke, 2021](https://arxiv.org/pdf/2105.13245.pdf) ?
They propose a constrained KG acquisition with a MC formulation and show it outperforms the "penalised KG" from [Chen et al., 2021](https://arxiv.org/abs/2101.08743).

**Comment by sdaulton** [comment]
> qKnowledgeGradient can run on ConstrainedMCObjective. Is there a reason for why doing so is not recommended in principle?

That should work, but I believe has not been thoroughly evaluated. For one, it seems KG may no longer be non-negative when there are constraints since evaluating a point that is infeasible seems like it could lower the feasible posterior maximizer). Empirically, this could work well though.

> what is your opinion on [Ungredda and Branke, 2021](https://arxiv.org/pdf/2105.13245.pdf)

I hadn't seen this before, but it looks interesting.

**Reply by polivucci** [reply]
> I hadn't seen this before, but it looks interesting.

I just noticed it has just come out [in print](https://dl.acm.org/doi/10.1145/3641544).
Table 1 also looks quite interesting to me:)

**Comment by slishak-PX** [comment]
Sorry to resurrect this old discussion.

Am I right in thinking that `ConstrainedMCObjective` has one advantage over `SampleReducingMCAcquisitionFunction`, in that it allows `best_f` for EI acquisition functions to be set more precisely?

For example, in https://botorch.org/tutorials/closed_loop_botorch_only, we set `best_f` to the best feasible observation so far. However, what if there are no feasible observations? Or what if we have an equality constraint (which can never be exactly satisfied)? 

With `ConstrainedMCObjective`, we can at least set `best_f` to the feasibility-weighted objective function which is being maximised. With `SampleReducingMCAcquisitionFunction`, we can keep randomly sampling until we find a feasible value, or we can guess a feasible `best_f` value, or we can just choose an infeasible `best_f`.

**Reply by slishak-PX** [reply]
@sdaulton not sure if you've seen this, but I'd appreciate your input on this!

**Reply by SebastianAment** [reply]
Hi @slishak-PX,

> Am I right in thinking that `ConstrainedMCObjective` has one advantage over `SampleReducingMCAcquisitionFunction`, in that it allows best_f for EI acquisition functions to be set more precisely?

In short: there is no difference in flexibility in setting `best_f`, and BoTorch uses the same value for both by default, see the following for details.

BoTorch uses [`compute_best_feasible_objective`](https://github.com/pytorch/botorch/blob/8b13120176a83080dc023cf85a1d33311e6e1cae/botorch/acquisition/utils.py#L92) to set the `best_f` value, e.g. in the [input constructors](https://github.com/pytorch/botorch/blob/8b13120176a83080dc023cf85a1d33311e6e1cae/botorch/acquisition/input_constructors.py#L1680C12-L1680C43), the [acquisition factories](https://github.com/pytorch/botorch/blob/8b13120176a83080dc023cf85a1d33311e6e1cae/botorch/acquisition/factory.py#L111), and [in the forward pass of any noisy EI variant](https://github.com/pytorch/botorch/blob/8b13120176a83080dc023cf85a1d33311e6e1cae/botorch/acquisition/logei.py#L436).

If there are no feasible observations, `compute_best_feasible_objective` estimates a lower bound on the unconstrained objective function via [`_estimate_objective_lower_bound`](https://github.com/pytorch/botorch/blob/8b13120176a83080dc023cf85a1d33311e6e1cae/botorch/acquisition/utils.py#L153C28-L153C59), which calls [`get_infeasible_cost`](https://github.com/pytorch/botorch/blob/8b13120176a83080dc023cf85a1d33311e6e1cae/botorch/acquisition/utils.py#L205), the same quantity that is used to specify the value of infeasible observations for `ConstrainedMCObjectives`, see e.g. [here in the input constructor of `qUCB`](https://github.com/pytorch/botorch/blob/8b13120176a83080dc023cf85a1d33311e6e1cae/botorch/acquisition/input_constructors.py#L820).

As a result, the `best_f` value will be identical for both set-ups if there are no feasible observations. Generally, there's no difference in flexibility in assigning `best_f` values, and you could implement your own version of `compute_best_feasible_objective` if you wanted more flexibility.

Since there are significant performance advantages of `SampleReducingAcquisitionFunctions` over `ConstrainedMCObjectives` for constrained problems, we strongly recommend the former.

> For example, in https://botorch.org/tutorials/closed_loop_botorch_only, we set best_f to the best feasible observation so far.

We should update this to use `compute_best_feasible_objective` for clarity, the current setup in the tutorial returns zero when there are no feasible observations which is a lower bound to "negative Hartmann", so this is in principle the same idea to inferring a lower bound based on the surrogate.

> Or what if we have an equality constraint (which can never be exactly satisfied)?

BoTorch doesn't support equality constraints for *outcomes*. Generally this doesn't seem to be applicable to real-world experiments, where measurement noise affects the outcomes, and it will give rise to the same problem for both `SampleReducingAcquisitionFunctions` and `ConstrainedMCObjectives` since they would both assign `best_f` based on the infeasible value (lower bound to the objective) computed by `get_infeasible_cost`. If there's a non-zero tolerance, one could use an upper and lower bound constraint that are apart by the tolerance value, so that feasibility is achievable. Exact constraints are more common on the inputs of the function, and can be incorporated by passing the [constraints to `optimize_acqf`](https://github.com/pytorch/botorch/blob/8b13120176a83080dc023cf85a1d33311e6e1cae/botorch/optim/optimize.py#L414).

> With `SampleReducingMCAcquisitionFunction`, we can keep randomly sampling until we find a feasible value, or we can guess a feasible `best_f` value, or we can just choose an infeasible `best_f`.

You don't need to do random sampling, and can use `SampleReducingMCAcquisitionFunction` whenever you can use `ConstrainedMCObjectives`.

Let me know if you have additional thoughts or questions on this.

**Reply by slishak-PX** [reply]
Thank you, this is really helpful. I was unaware of much of the functionality you highlighted - it would be great if some of it was documented/used in some of the tutorials rather than buried in the API documentation!

I take your point on equality constraints not being applicable on real-world noisy observations, but not all applications of BO involve significant noise: if your black box function is deterministic yet expensive and nondifferentiable then you can imagine settings where you might want to set equality constraints on your outputs. As you say, as soon as you apply a tolerance, you of course just end up with a pair of inequality constraints.

**Reply by SebastianAment** [reply]
> it would be great if some of it was documented/used in some of the tutorials rather than buried in the API documentation!

Yes we'll work on this, thanks for the feedback.

> want to set equality constraints on your outputs

You can of course also achieve this with two inequality constraints with the same bound using the current functionality. If you are using any member of the `LogEI` family of acquisition functions (e.g. [`qLogNoisyExpectedImprovement`](https://github.com/pytorch/botorch/blob/cd657d9c5fa47461e152bbfbcdf84f068b1141b8/botorch/acquisition/logei.py#L239) or [`qLogNoisyExpectedHypervolumeImprovement`](https://github.com/pytorch/botorch/blob/cd657d9c5fa47461e152bbfbcdf84f068b1141b8/botorch/acquisition/multi_objective/logei.py#L319)), this could still work because the optimizer would get gradient information from the constraint component, even if most of the search space is infeasible. 

This would make the policy aggressively search for feasibility, first and foremost. If you are ok with even a small deviation from the exact equality constraint, it would probably give you faster improvements in the objective function. 

If you do end up trying this out, I'd be curious to hear about your experience!

**Comment by LXlearning** [comment]
I would like to know how to handle situations with multiple constraints. Currently, in qLogNoisyExpectedImprovement, it seems that I can only construct such a function for passing in constraints as follows:
```
constraints = [lambda Z: Z[..., problem_num_objs:].amax(dim=-1)]
acq_func = qLogExpectedHypervolumeImprovement(
        model=model,
        ref_point=ref_point,
        partitioning=partitioning,
        sampler=sampler,
        objective=IdentityMCMultiOutputObjective(outcomes=list(range(problem_num_objs))),
        constraints=constraints,
    )
```
However, this way of handling is too crude. I want to use this formula to calculate the feasibility of the constraints: 

 $`\alpha_{EIC}(x|f_{c}^{*})=\alpha_{EI}(x|f_{c}^{*})\prod_{j = 1}^{J}\mathbb{P}(c_{j}(x)\leq0)`$
 But it seems that there is no interface that allows me to do so, How can I add the constraints parameter in qLogExpectedHypervolumeImprovement to implement the feasibility probability scheme for multiple constraints?

**Reply by SebastianAment** [reply]
Hi @LXlearning, 

The interface is designed with this in mind, you can just pass a list of constraints:
```
constraints = [(lambda Z, i=i: Z[..., num_objectives + i]) for i in range(num_constraints)]
```
Note that the `i=i` is necessary for the closure to capture the variable correctly. 

Passing this to any improvement-based acquisition function will weigh the (log) improvement by the (log) probability of feasibility of all constraints as you intend.

Let me know if you have further questions about this.

**Reply by LXlearning** [reply]
Thank you very much. This is of great help to me.

I want to add explicit constraints to my optimization problem. The input parameter of my optimization problem is a dictionary, like this:
```
X = {'x1': 1, 'x2':1}
# This is a explicit constraint(Unlike implicit constraints, it does not need to be calculated through the fitting of a black-box model）
# This is a simple linear problem, but it may actually be non-linear.
def explicit_constrains_test(X):
    return  X['x1'] + X['x2'] <= 10
```

I noticed that both `optimize_acqf` and `optimize_acqf_list` support the input parameter `nonlinear_inequality_constraints`, but I get an error when I pass it in.

```
RuntimeError: `ic_generator` must be given if there are non-linear inequality constraints.
```

I would like to ask how the two parameters `ic_generator` and `nonlinear_inequality_constraints` should be passed in.

**Reply by SebastianAment** [reply]
Hi @LXlearning, 

On a higher level, I encourage you to to read the documentation about these arguments e.g. [`nonlinear_inequality_constraints`](https://github.com/pytorch/botorch/blob/299438bb1c6f9b292db6e8a9cd5c5acf608c0968/botorch/optim/optimize.py#L531C9-L531C41) or the outcome [`constraints`](https://github.com/pytorch/botorch/blob/299438bb1c6f9b292db6e8a9cd5c5acf608c0968/botorch/acquisition/monte_carlo.py#L221C1-L222C1) before. Let us know if something isn't clear about the explanations there.

Note that the argument to the non-linear input constraint function will always be a tensor, since it is called within BoTorch, and needs to return a value that is constrained to be positive ([as indicated by the documentation](https://github.com/pytorch/botorch/blob/299438bb1c6f9b292db6e8a9cd5c5acf608c0968/botorch/optim/optimize.py#L533C65-L533C66)), not a Boolean, identical to the outcome constraints specified above, e.g. `X.sum(dim=-1) - 10` for the example you shared.

For an example of a `ic_generator` and its interface, see the default [`get_batch_initial_conditions` ](https://github.com/pytorch/botorch/blob/299438bb1c6f9b292db6e8a9cd5c5acf608c0968/botorch/optim/initializers.py#L245). This is necessary to generate feasible starting points, which requires further knowledge of the constraints in the non-linear case.

**Reply by LXlearning** [reply]
Thank you. I wrote the following code with reference to the help documentation.
```
def feasible_func_botorch(X):
    X_ori = unnormalize(X.detach(), bounds=x_bounds)
    return X_ori.sum(dim=-1) - 10
nonlinear_inequality_constraints = [(feasible_func_botorch2, True)]
candidates, _ = optimize_acqf(acq_function=acq_func,
                                                    bounds=standard_bounds,
                                                    q=BATCH_SIZE,
                                                    num_restarts=NUM_RESTARTS,
                                                    raw_samples=RAW_SAMPLES,  # used for intialization heuristic
                                                    options={"batch_limit": 1, "maxiter": 100},
                                                    nonlinear_inequality_constraints=nonlinear_inequality_constraints,
                                                    ic_generator=gen_batch_initial_conditions,
                                                    sequential=True,
        )
```
I used this constraint in the optimize_acqf function, but encountered two new problems:

1. Since the `make_scipy_nonlinear_inequality_constraints` function stipulates that `batch_initial_conditions` must satisfy the non-linear inequality, if I define a relatively strict constraint, how can I ensure that the initial sampling satisfies my non-linear inequality?

2. Because the design variables were normalized during the model training of BO, it is necessary to unnormalize them when calculating the constraints here. However, after I defined the function in this way, it would cause the function to be unable to calculate the gradient information and result in an optimization error. How can this be solved? 

```
element 0 of tensors does not require grad and does not have a grad_fn
  File "/Users/luoxin/miniforge3/envs/lx/lib/python3.9/site-packages/torch/autograd/__init__.py", line 394, in grad
    result = Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  File "/Users/luoxin/miniforge3/envs/lx/lib/python3.9/site-packages/botorch/generation/gen.py", line 211, in f_np_wrapper
    gradf = _arrayify(torch.autograd.grad(loss, X)[0].contiguous().view(-1))
```

**Reply by LXlearning** [reply]
@SebastianAment 
There are some cases where the constraint function is inexpensive but also a black box and the gradient cannot be directly computed. An error will occur due to the gradient problem when defining `nonlinear_inequality_constraints` in `optimize_acqf` Can you tell me how to configure `nonlinear_inequality_constraints` with some  black box constraint function?

One method I can think of is to sample a large number of sample points and fit these constraints using the response surface method to obtain the gradients. However, this is very inefficient and not very convenient when dealing with high-dimensional problems.

Looking forward to your answer. Thank you very much.

**Reply by SebastianAment** [reply]
Hi @LXlearning, 

BoTorch supports black box constraints by passing the `constraints` argument to the acquisition function, as you wrote about initially. The `nonlinear_inequality_constraints` only support known differentiable constraints. 

> constraint function is inexpensive

If sampling the constraint is much less costly than the objective, fitting a deterministic surrogate to a lot of samples from the constraint function before starting the optimization could make sense. 

If this is not efficient enough, the constraint can be modeled as a GP too, and the acquisition function will balance exploration and exploitation of both the objective and feasibility space in a sample-efficient way.

**Reply by LXlearning** [reply]
Thank you for the answer.
But building a GP model for cheap display constraints may be relatively inefficient, especially in high-dimensional situations.
I think using an evolutionary algorithm to optimize the acquisition function with display constraints (replacing the current gradient algorithm) may also be an effective means. I hope that botorch can also provide this method in the future.
